### PR TITLE
fix: Set MSRV for internal packages

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

Correctly communicates the MSRV we support to our users.

For packages that are more generally mean to be used by other people, I've punted on for now (see ehuss' comment on this PR).  We'll likely need to figure this out for #12432 though

### Additional information

This is prep for a future change which will have us use a fixed rust version for the semver tests with a PR updating them.

### CODE REVIEW

**Review by weihanglo — COMMENTED** [review]
It is technically true that Cargo-the-library is only guaranteed to be compatible with one corresponding Rust version. And this also gets rid off some burdens we would have on MSRV. However, it is a bit too strict for every crate to stick to the latest stable. Someone may only depends on a tiny part of cargo-the-library but with this MSRV bump that couldn't anymore.

**Comment by weihanglo** [comment]
> This is prep for a future change which will have us use a fixed rust version for the semver tests with a PR updating them.

What will this look like? I didn't get the full picture maybe because of missing this?

**Comment by epage** [comment]
>  Someone may only depends on a tiny part of cargo-the-library but with this MSRV bump that couldn't anymore.

Unless its behind a feature flag or relying on a behavior change, MSRV is all-or-nothing.

**Comment by weihanglo** [comment]
Some crates like `crates-io` and `home` are pretty not directly coupled with `cargo`, feel like they shouldn't be set with the same MSRV.

(and perhaps `cargo-util` if we have a firmed definition of what utils are…)

**Comment by epage** [comment]
> Some crates like crates-io and home are pretty not directly coupled with cargo, feel like they shouldn't be set with the same MSRV.

The choices I'm aware of
- Don't set MSRV and put it in users hands
  - Short term: users have to play around to find the version that works
  - Long term: with MSRV aware resolver, users will still have to play around to find the version that works
- Set to latest
  - Might be too high in some cases
  - Easy to verify
- Set different MSRVs
  - To ensure accuracy, we'd need to add MSRV check jobs
  - We'd need to hand-manage the list of which packages to verify with which MSRV or write a tool
- Manage multiple workspaces, one of the core of cargo and one for more external stuff
  -This might be of more interest if we start taking responsibility for Rust APIs for cargo (e.g. build scripts, metadata)

Considering
- `home` is primarily for cargo (#12386) (and `cargo-util` looks similar)
- So far, we've not been looking to take on the maintenance load of clients for our crates (making them more "buyer beware")
- We want our process to scale with adding more packages, including CI

To me, it seems like its a choice to no MSRV or latest (until / unless we decide to have multiple workspaces) and setting it would be useful so I would lean towards that.

Thoughts?

**Comment by weihanglo** [comment]
BTW, version bump check is somehow broken. See <https://github.com/rust-lang/cargo/issues/12347>. You may need to manually check if it does need a bump for nightly.

(I'll try fixing that in the next few days)

**Comment by ehuss** [comment]
I'm going to check my box, though there were some mild concerns that I think were already highlighted above, but I wanted to repeat back to make sure they are understood.

- There is a desire to make it easy to add new packages, but if we had independent MSRV for different packages, that might require some manual management to make sure they are set and tested in CI correctly.
- Adding new syntax to our workspace `Cargo.toml` that wasn't forwards compatible would force us to hold on making that change until that support hit the MSRV version. For example, if we used something like the recently added named-debug options, we would have to hold off on dogfooding it for a long while.
- Something like `build-rs` (#12432) would need to be mindful of which versions of cargo it supports. For example, if we had a package like that, and we switched it to use the new `cargo::` syntax, then we could only make that change once that support passes the MSRV.

Some possible alternate solutions:
- Multiple workspaces.
- Have an xtask to help manage things.
- Allow commands like `cargo check` to skip unsupported rust versions (either built-in to cargo, or an xtask like thing). This wouldn't help with manifest syntax changes.
- Don't set MSRV at all for now.
- Set MSRV, but not for every package (like the credential providers).

My concerns were mostly around the credential providers, cargo-platform, and home. For now, credential providers are unstable, so the exact MSRV isn't important. But I think once we head towards stabilization, we should try to figure out a solution. I think it will be important to make sure that people can `cargo install` providers with some MSRV older than latest stable.

`cargo-platform` almost never changes, and it will be stable-2 when it publishes, which is the MSRV I would suggest anyway. The only concern is if we need to do an out-of-band publish, which I don't remember if we've ever done, but would be very exceptional. That seems like something that we can improve later.

`home` is in a weird state, but I don't think we will be releasing a new version in the short term. (There are some unpublished version bumps, but I don't know what's going on there.) This also seems like something we can figure out, but we need to be mindful of.

I would have a slight preference towards not setting the MSRV in the short term for the things that might cause problems (like cargo-platform or home), just so that this doesn't introduce new pain points.

**Comment by epage** [comment]
>  but I wanted to repeat back to make sure they are understood.
>
> - Adding new syntax to our workspace Cargo.toml that wasn't forwards compatible would force us to hold on making that change until that support hit the MSRV version.

There is also the issue that shared dependencies need to meet the lowest common denominator of the MSRVs.

**Comment by epage** [comment]
I've gone ahead and limited this to more internal scripts, punting on how to handle MSRV testing for more public packages to #12432

**Comment by epage** [comment]
@Turbo87 mind double checking my RenovateBot regex manager?

File: .github/renovate.json5 [hunk 1]
```diff
@@ -8,7 +8,32 @@
   ignorePaths: [
     "**/tests/**",
   ],
+  regexManagers: [
+    {
+      fileMatch: [
+        '^Cargo.toml$',
+      ],
+      matchStrings: [
+        'rust-version.*?(?<currentValue>\\d+\\.\\d+(\\.\\d+)?)',
```
Comment (Turbo87) [inline .github/renovate.json5:17]:
```suggestion
        'rust-version\\s*=\\s*"(?<currentValue>\\d+\\.\\d+(\\.\\d+)?)"',
```

I would probably go with this slightly more specific regex, but I think it should work either way :)

I'm using something roughly similar in https://github.com/Turbo87/renovate-config/blob/04f6fb54cca78973deacec582fb36b4bf5f2da9c/rust/updateToolchain.json#L8

**Review by weihanglo — APPROVED** [review]

File: .github/renovate.json5 [hunk 2]
```diff
@@ -8,7 +8,33 @@
...
+      customType: 'regex',
+      fileMatch: [
+        '^Cargo.toml$',
+      ],
+      matchStrings: [
+        'rust-version.*?(?<currentValue>\\d+\\.\\d+(\\.\\d+)?)',
+      ],
+      depNameTemplate: 'latest-msrv',
+      packageNameTemplate: 'rust-lang/rust',
+      datasourceTemplate: 'github-releases',
```
Comment (weihanglo) [inline .github/renovate.json5:22]: Just mind that `1.71.1` is not yet in [GitHub Release](https://github.com/rust-lang/rust/releases).

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Just realized, with this pull request and #12395, it is inevitable that every time a Rust release comes out, `cargo-util` and `crates-io` will be forced to release a version due to the bump of `package.rust-version`, even when there is no actual code change involved.

Is that something we've considered? It feels like a bit unnecessary.

**Pull request merged**